pkg/targets/reconciler/cloudeventstarget/adapter.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudeventstarget
+
+import (
+	"path"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"knative.dev/eventing/pkg/reconciler/source"
+	"knative.dev/pkg/apis"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+
+	commonv1alpha1 "github.com/triggermesh/triggermesh/pkg/apis/common/v1alpha1"
+	"github.com/triggermesh/triggermesh/pkg/apis/targets/v1alpha1"
+	common "github.com/triggermesh/triggermesh/pkg/reconciler"
+	"github.com/triggermesh/triggermesh/pkg/reconciler/resource"
+)
+
+const (
+	envCloudEventsPath                  = "CLOUDEVENTS_PATH"
+	envCloudEventsURL                   = "CLOUDEVENTS_URL"
+	envCloudEventsBasicAuthUsername     = "CLOUDEVENTS_BASICAUTH_USERNAME"
+	envCloudEventsBasicAuthPasswordPath = "CLOUDEVENTS_BASICAUTH_PASSWORD_PATH"
+)
+
+// adapterConfig contains properties used to configure the target's adapter.
+// Public fields are automatically populated by envconfig.
+type adapterConfig struct {
+	// Configuration accessor for logging/metrics/tracing
+	obsConfig source.ConfigAccessor
+	// Container image
+	Image string `default:"gcr.io/triggermesh/cloudeventstarget-adapter"`
+}
+
+// Verify that Reconciler implements common.AdapterServiceBuilder.
+var _ common.AdapterServiceBuilder = (*Reconciler)(nil)
+
+// BuildAdapter implements common.AdapterServiceBuilder.
+func (r *Reconciler) BuildAdapter(trg commonv1alpha1.Reconcilable, _ *apis.URL) (*servingv1.Service, error) {
+	typedTrg := trg.(*v1alpha1.CloudEventsTarget)
+
+	options := []resource.ObjectOption{
+		resource.Image(r.adapterCfg.Image),
+		resource.EnvVars(makeAppEnv(typedTrg)...),
+		resource.EnvVars(r.adapterCfg.obsConfig.ToEnvVars()...),
+	}
+
+	if typedTrg.Spec.Credentials != nil {
+		secretName := "basicauths"
+		secretPath := "/opt/basicauths"
+		secretFileName := "cesource"
+
+		options = append(options, resource.EnvVar(envCloudEventsBasicAuthUsername, typedTrg.Spec.Credentials.BasicAuth.Username))
```

Comment:
A word of warning when adding optional env vars at the **end** of the container's env list.

I know I've been pedantic by refactoring a few components to move custom env vars upper in the env list (including the CE source recently), but I've done that for a good reason: _deep derivative comparisons_.

If revision 00001 has:

```yaml
env:
- name: CLOUDEVENTS_URL
  value: https://pablo.example.com
- name: K_LOGGING
  value: {...}
- name: CLOUDEVENTS_BASICAUTH_USERNAME
  value: alice
```

then a user disables basic auth, the next revision will compare to this:

```diff
 env:
 - name: CLOUDEVENTS_URL
   value: https://pablo.example.com
 - name: K_LOGGING
   value: {...}
-- name: CLOUDEVENTS_BASICAUTH_USERNAME
-  value: alice
```

and in a deep derivative comparison, extra values appended at the _end_ of a list are ignored (to allow e.g. webhooks to inject stuff like annotations, env vars, etc. without causing reconciliation conflicts).

If, however, fixed generated values such as K_LOGGING remain at the end of the list, the comparison will correctly return `false` because slice indices will be shifted by one:

```diff
 env:
 - name: CLOUDEVENTS_URL
   value: https://pablo.example.com
-- name: CLOUDEVENTS_BASICAUTH_USERNAME
-  value: alice
 - name: K_LOGGING
   value: {...}
```

It's mostly a trick, but potentially an important one to consider to prevent users from being surprised by small bugs like this.
Also, it's not only valid for env vars but for any slice type, but env vars are a common pitfall in most of our implementations.

---

Pushing env vars in the following order solves the issue I just described:

https://github.com/triggermesh/triggermesh/blob/8102ddd1a97813d0254b898cd07992a129fa313c/pkg/sources/reconciler/cloudeventssource/adapter.go#L114-L126

Reply:
totally forgot about this.

https://github.com/triggermesh/triggermesh/pull/829/commits/f4dca56c960ee78f308d5d83ce1f132b7eafe150